libraries/botbuilder-dialogs-adaptive/src/conditions/onCondition.ts

```diff
@@ -5,21 +5,121 @@
+
+    private _extraConstraints: Expression[] = [];
+    private _fullConstraint: Expression = null;
+
+
+    /**
+     * Gets or sets the condition which needs to be met for the actions to be executed (OPTIONAL).
+     */
+    public condition: string;
+
+    /**
+     * Gets or sets the actions to add to the plan when the rule constraints are met.
+     */
+    public actions: Dialog[] = [];
+
+    /**
+     * Create a new `OnCondition` instance.
+     * @param condition (Optional) The condition which needs to be met for the actions to be executed.
+     * @param actions (Optional) The actions to add to the plan when the rule constraints are met.
+     */
+    constructor(condition: string = null, actions: Dialog[] = []) {
+        this.condition = condition;
+        this.actions = actions;
+    }
+
+    /**
+     * Get the expression for this condition
+     * @param parser Expression parser.
+     * @returns Expression which will be cached and used to evaluate this condition.
+     */
+    public getExpression(parser: ExpressionParserInterface): Expression {
+        if (this._fullConstraint == null) {
+            const allExpressions: Expression[] = [];
+            if (this.condition) {
+                try {
+                    allExpressions.push(parser.parse(this.condition));
+                } catch (err) {
+                    throw Error(`Invalid constraint expression: ${this.condition}, ${err.toString()}`);
+                }
+            }
+
+            if (this._extraConstraints.length > 0) {
+                allExpressions.push(...this._extraConstraints);
+            }
+
+            if (allExpressions.length > 1) {
+                return Expression.makeExpression(ExpressionType.And, undefined, ...allExpressions);
+            } else if (allExpressions.length == 1) {
+                return allExpressions[0];
+            } else {
+                return new Constant(true);
+            }
+        }
+
+        return this._fullConstraint;
+    };
+
+    /**
+     * Add external condition to the OnCondition
+     * @param condition External constraint to add, it will be AND'ed to all other constraints.
+     */
+    public addExternalCondition(condition: string) {
+        if (condition) {
+            try {
+                const parser = new ExpressionEngine();
+                this._extraConstraints.push(parser.parse(condition));
+                this._fullConstraint = null;
+            } catch (err) {
+                throw Error(`Invalid constraint expression: ${condition}, ${err.toString()}`);
+            }
+        }
+    }
+
+    /**
+     * Method called to execute the condition's actions.
+     * @param planningContext Context.
+     * @returns A promise with plan change list.
+     */
+    public async executeAsync(planningContext: SequenceContext): Promise<ActionChangeList[]> {
+        return Promise.resolve([this.onCreateChangeList(planningContext)]);
```

Comment:
>Promise.resolve( [](start = 15, length = 16)

you don't need this if you're using the async keyword... you can just return the array list.